File: backend/backend/src/main/java/Shapes/Shape.java
package Shapes;

import HelpingClasses.Position;
import HelpingClasses.Scale;
import System.MyCloneable;
import System.Convertable;
import org.json.simple.JSONObject;


public abstract class Shape implements MyCloneable, Convertable
{
    private int ID;
    private boolean Enabled;
    private double StrokeWidth;
    private String StrokeColor;
    private Position ShapePosition;
    private Scale MyScale;

    @Override
    public void Clone(Shape shape)
    {
        shape.SetStrokeColor(this.StrokeColor);
        shape.SetStrokeWidth(this.StrokeWidth);
        shape.Enabled = this.Enabled;
        shape.SetScale(this.MyScale);
        shape.SetPosition(this.ShapePosition);
    }

    @Override
    public void PutObjectInJson(JSONObject jsonObject)
    {
        jsonObject.put("stroke", this.StrokeColor);
        jsonObject.put("strokewidth", this.StrokeWidth);
        jsonObject.put("Enabled", this.Enabled);
        this.MyScale.PutObjectInJson(jsonObject);
        this.ShapePosition.PutObjectInJson(jsonObject);
    }

    public void SetScale(Scale scale)
    {
        this.MyScale = scale;
    }

    public Scale GetScale()
    {
        return this.MyScale;
    }

    public int GetId()
    {
        return this.ID;
    }
    public String GetStrokeColor()
    {
        return this.StrokeColor;
    }
    public void SetStrokeColor(String color)
    {
        this.StrokeColor = color;
    }
    public double GetStrokeWidth()
    {
        return this.StrokeWidth;
    }

    public void SetStrokeWidth(double width)
    {
        this.StrokeWidth = width;
    }
    public void SetEnabled()
    {
        this.Enabled = true;
    }
    public void SetDisabled()
    {
        this.Enabled = false;
    }
    public void SetPosition(Position position)
    {
        this.ShapePosition = position;
    }
    public Position GetPosition()
    {
        return this.ShapePosition;
    }
    public Shape(int id)
    {
        this.ID = id;
    }
}

File: frontend/paint_frontend/src/app/classes/Shape.ts
export abstract class Shape{
    strokeWidth: number;
    id: number;
    enabled: boolean;
    strokeColor: string;

    constructor( id:number, strokeWidth?:number, enabled?:boolean, strokeColor?:string){
        this.id = id;
        this.strokeWidth = strokeWidth || 1;
        this.enabled = enabled || true;
        this.strokeColor = strokeColor || "FFFFFF";
    }

    initialize(jsonObject:any):void{
        this.strokeWidth = jsonObject.strokeWidth;
        this.id = jsonObject.id;
        this.enabled = jsonObject.enabled;
        this.strokeColor = jsonObject.strokeColor;
    }

    getStrokeColor():string{
        return this.strokeColor;
    }

    setStrokeColor(strokeColor:string):void{
        this.strokeColor = strokeColor;
    }

    getStrokeWidth():number{
        return this.strokeWidth;
    }

    setStrokeWidth(strokeWidth:number):void{
        this.strokeWidth = strokeWidth;
    }

    getId():number{
        return this.id;
    }

    setEnabled():void{
        //enables object
        this.enabled = true;
    }

    setDisabled():void{
        //disables object
        this.enabled = false;
    }

    draw(layer:any):void{}

    

}